diff --git a/src/Exception/ValidationException.php b/src/Exception/ValidationException.php
index 7a6fb99..a19c989 100644
--- a/src/Exception/ValidationException.php
+++ b/src/Exception/ValidationException.php
@@ -33,7 +33,8 @@ class ValidationException extends FeedsRuntimeException {
    */
   public function __construct($message = '', array $args = []) {
     $this->messageString = $message;
-    $this->message = new FormattableMarkup($message, $args);
+    $formattedMarkup = new FormattableMarkup($message, $args);
+    $this->message = $formattedMarkup->__toString();
     $this->args = $args;
     $this->code = 0;
   }
