diff --git a/views_infinite_scroll.module b/views_infinite_scroll.module
index 290e446..47f3e07 100644
--- a/views_infinite_scroll.module
+++ b/views_infinite_scroll.module
@@ -40,14 +40,9 @@ function views_infinite_scroll_preprocess_views_view(&$vars) {
   if ($view->getDisplay()->isPagerEnabled() && !empty($vars['rows'])) {
     $pager = $view->getPager();
     if ($pager && $pager instanceof InfiniteScroll) {
-      $vars['rows']['#theme_wrappers'] = ['container'];
-      $vars['rows']['#attributes'] = [
-        'data-drupal-views-infinite-scroll-content-wrapper' => TRUE,
-        'class' => [
-          'views-infinite-scroll-content-wrapper',
-          'clearfix',
-        ],
-      ];
+      $vars['rows']['#attributes']['data-drupal-views-infinite-scroll-content-wrapper'] = TRUE;
+      $vars['rows']['#attributes']['class'][] = 'views-infinite-scroll-content-wrapper';
+      $vars['rows']['#attributes']['class'][] = 'clearfix';
     }
   }
 }
