diff --git a/core/misc/ajax.js b/core/misc/ajax.js
index a899f4a9b9..d0888b1a93 100644
--- a/core/misc/ajax.js
+++ b/core/misc/ajax.js
@@ -179,6 +179,7 @@ function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr
     }
 
     if (!this.url) {
+      this.url = '';
       var $element = $(this.element);
       if ($element.is('a')) {
         this.url = $element.attr('href');
