diff --git a/src/Plugin/Block/EntityEditFormBlock.php b/src/Plugin/Block/EntityEditFormBlock.php
index cdd6a3c..ce92a02 100644
--- a/src/Plugin/Block/EntityEditFormBlock.php
+++ b/src/Plugin/Block/EntityEditFormBlock.php
@@ -30,6 +30,7 @@ class EntityEditFormBlock extends BlockBase {
     return array(
       'entity_type' => '',
       'bundle' => '',
+      'form_mode' => '',
     );
   }
 
@@ -82,6 +83,13 @@ class EntityEditFormBlock extends BlockBase {
       '#default_value' => $this->configuration['entity_type'] . '.' . $this->configuration['bundle'],
     );
 
+    $form['form_mode'] = array(
+      '#title' => $this->t('Form Mode'),
+      '#type' => 'textfield',
+      '#required' => TRUE,
+      '#default_value' => $this->configuration['form_mode'],
+    );
+
     return $form;
   }
 
@@ -93,6 +101,7 @@ class EntityEditFormBlock extends BlockBase {
     $values = explode('.', $selected_entity_type_bundle);
     $this->configuration['entity_type'] = $values[0];
     $this->configuration['bundle'] = $values[1];
+    $this->configuration['form_mode'] = $form_state->getValue('form_mode');
   }
 
   /**
@@ -114,8 +123,10 @@ class EntityEditFormBlock extends BlockBase {
       $entity->setOwnerId(\Drupal::currentUser()->id());
     }
 
+    $form_mode = !empty($this->configuration['form_mode']) ? $this->configuration['form_mode'] : 'default';
+
     $form = \Drupal::entityManager()
-      ->getFormObject($this->configuration['entity_type'], 'default')
+      ->getFormObject($this->configuration['entity_type'], $form_mode)
       ->setEntity($entity);
     return \Drupal::formBuilder()->getForm($form);
   }
